airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/cursor_based/MysqlCursorBasedUtils.java

```diff
@@ -0,0 +1,107 @@
+        }
+
+        final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(streamDescriptor.getName(),
+                                                                                       streamDescriptor.getNamespace());
+
+        // Build a map of stream <-> initial load status for streams that currently have an initial primary
+        // key load in progress.
+
+        if (streamState.has(STATE_TYPE_KEY)) {
+          if (streamState.get(STATE_TYPE_KEY).asText().equalsIgnoreCase(PRIMARY_KEY_STATE_TYPE)) {
+            final PrimaryKeyLoadStatus primaryKeyLoadStatus = Jsons.object(streamState, PrimaryKeyLoadStatus.class);
+            pairToInitialLoadStatus.put(pair, primaryKeyLoadStatus);
+            streamsStillInPkSync.add(pair);
+          } else if (streamState.get(STATE_TYPE_KEY).asText().equalsIgnoreCase(CURSOR_BASED_STATE_TYPE)) {
+            cursorBasedStreamPairs.add(pair);
+          }
+        }
+        alreadySeenStreamPairs.add(new AirbyteStreamNameNamespacePair(streamDescriptor.getName(), streamDescriptor.getNamespace()));
+      });
+    }
+    final List<ConfiguredAirbyteStream> streamsForPkSync = new ArrayList<>();
+    final List<ConfiguredAirbyteStream> streamsForCursorBasedSync = new ArrayList<>();
+    fullCatalog.getStreams().stream()
+        .filter(stream -> {
+          System.out.println(stream.getStream().getSourceDefinedPrimaryKey().toString());
+          return streamsStillInPkSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream()))
+              && streamHasPrimaryKey(stream);
+        })
+        .map(Jsons::clone)
+        .forEach(streamsForPkSync::add);
+
+    fullCatalog.getStreams().stream()
+        .filter(stream -> cursorBasedStreamPairs.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream()))).map(Jsons::clone)
+        .forEach(streamsForCursorBasedSync::add);
+    final List<ConfiguredAirbyteStream> newlyAddedStreams = identifyStreamsToSnapshot(fullCatalog, Collections.unmodifiableSet(alreadySeenStreamPairs));
+    streamsForPkSync.addAll(newlyAddedStreams);
+    final Map<AirbyteStreamNameNamespacePair, CursorBasedStatus> pairToCursorBasedStatus =
+        getCursorBasedSyncStatusForStreams(database, streamsForPkSync, stateManager, quoteString);
+
+    return new StreamsCategorised(new InitialLoadStreams(streamsForPkSync, pairToInitialLoadStatus),
+                                  new CursorBasedStreams(streamsForCursorBasedSync, pairToCursorBasedStatus));
+  }
+
+  private static boolean streamHasPrimaryKey(final ConfiguredAirbyteStream stream) {
+    return stream.getStream().getSourceDefinedPrimaryKey().size() > 0;
+  }
+
+  public record CursorBasedStreams(List<ConfiguredAirbyteStream> streamsForCursorBased,
+                                   Map<AirbyteStreamNameNamespacePair, CursorBasedStatus> pairToCursorBasedStatus) {
+
+  }
+
+  public record StreamsCategorised(InitialLoadStreams initialLoadStreams,
```

Comment:
can we remove this and just keep InitialLoadStreams? Anything that is not in InitialLoadStreams would implicitly be a cursor based streams right?

Reply:
Yep! Addressed above.